adamacs/ingest/pyrat.py

```diff
@@ -35,20 +72,142 @@ def __init__(self):
+                                              'description': mutation['mutationname']})
+        _ = [d.pop('mut_line') for d in mutation_list]  # drop unique ID
+
+        print('Gathering subjects...')
+        subject_list = []
+        genotype_list = []
+        for animal in restrict_by(payload, 'eartag_or_id', subject.Subject, 'subject'):
+            parents = animal['parents']
+            if parents:
+                parent_ids = {parents[0]['parent_sex']: parents[0]['parent_eartag'],
+                              parents[1]['parent_sex']: parents[1]['parent_eartag']}
+            else:
+                parent_ids = None
+            subject_list.append({'subject': animal['eartag_or_id'],
+                                 'earmark': animal['labid'],
+                                 'sex': animal['sex'],
+                                 'birth_date': animal['dateborn'] or None,
+                                 # 'subject_description': ??, # pyrat animal_color?
+                                 'generation': animal['generation'] or None,
+                                 'parent_ids': parent_ids or None,
+                                 'owner_id': owner_id,
+                                 'responsible_id': responsible_id,
+                                 'line': animal['strain_id'],
+                                 'protocol': animal['licence_number'] or None})
+            for mutation in animal['mutations']:
+                genotype_list.append({'subject': animal['eartag_or_id'],
+                                      'line': animal['strain_id'],
+                                      'mutation_id': mutation['mutation_id'],
+                                      'genotype': mutation['mutationgrade']})
+
+        # --- User Confirmation ---
+        print('--- PyRAT items to be inserted ---')
+        inserts = [user_list, protocol_list, line_list, mutation_list, subject_list]
+        print_key = ['name', 'protocol_description', 'line_name', 'description',
+                     'subject']
+        titles = ['User(s)', 'Protocol(s)', 'Line(s)', 'Mutation(s)', 'Subjects']
+        for insert, key, title in zip(inserts, print_key, titles):
+            print(f'{title}: ', list(i[key] for i in insert), '\n')
+        if prompt and dj.utils.user_choice('Proceed with new subject schema insert?'
+                                           ) != 'yes':
+            print('Canceled insert.')
+            return
+
+        # --- Inserts ---
+        subject.User.insert(user_list)
+        subject.Protocol.insert(protocol_list)
+        subject.Line.insert(line_list)
+        # need additional logic above to prevent adding duplicates to list
+        subject.Mutation.insert(mutation_list)  # , skip_duplicates=True)
+        subject.Subject.insert(subject_list)
+        # foreign key contstraint fail bc line_list is missing items
+        subject.SubjectGenotype.insert(genotype_list)
```

Comment:
```suggestion
    with subject.Subject.connection.transaction:
                # --- Inserts ---
                subject.User.insert(user_list)
                subject.Protocol.insert(protocol_list)
                subject.Line.insert(line_list)
                # need additional logic above to prevent adding duplicates to list
                subject.Mutation.insert(mutation_list)  # , skip_duplicates=True)
                subject.Subject.insert(subject_list)
                # foreign key contstraint fail bc line_list is missing items
                subject.SubjectGenotype.insert(genotype_list)
```